### DISCUSSION TITLE
mod files not built (Windows11/vs2022/oneAPI2023/CMake 3.22.1)

### DISCUSSION THREAD

**Discussion by jalvesz** [Cambridge-ICCS/FTorch#74] [Q&A] [answered]
Hi, I'm trying to test the library but I'm hitting a small issue with the build system. I'm compiling under Windows 11 + Visual Studio 2022 + oneAPI 2023.2.1 using CMake 3.22.1. 

I included this library as a third_party_lib within my project. After launching my build scripts the full project is created and I can see ftorch within, running the build of ftorch finishes successfully BUT I only see the following files: ftorch.dll / ftorch.exp / ftorch.lib / ftorch.pdb within the Debug folder, can't find the modules folder with the .mod. 

Curiously, looking at the Ftorch folder in the VS project, it seems that Visual sees it as a C/C++ project only. Maybe this is why VS is not creating the .mod s files?:
![image](https://github.com/Cambridge-ICCS/FTorch/assets/102541118/675d129b-0c68-4035-888e-33dccf98995d)

Without the .mod files I'm unable to actually use it within the rest of my code, so I wanted to ask for any advise/hints in case you have seen such issue before? Thanks a lot!!

**Comment by jatkinson1000** [comment]
Hi @jalvesz Thanks for asking. How recently was it that you tried this? We definitely saw similar issues when working on windows builds, but updated the instructions late last week.
These are on the readme including:

>If building on Windows you will need to add -G "NMake Makefiles" to the cmake command. Also you will need to load the intel fortran compilers using setvars.bat ([see Intel docs](https://www.intel.com/content/www/us/en/docs/oneapi/programming-guide/2023-2/use-the-setvars-script-with-windows.html)) which is found in the Intel compiler install directory.

There is also issue #50 discussing this.

If you have tried with the most recent instructions and still get issues then we can look into this in more detail with you.

**Reply by jalvesz** [reply]
Hi @jatkinson1000 thanks for the prompt answer! So I think I know where the "problem" is coming from, I had forgotten that MSBuild doesn't support mixing Fortran and C/C++ https://community.intel.com/t5/Intel-Fortran-Compiler/C-Fortran-Mixed-language-Environment-in-Visual-Studio-2015/td-p/1160315 so one has two create to separate static libraries and link in the correct order accordingly. With NMake it is possible to build it, but if I'm not wrong, it will not generate the .sln in order to do intereactive debuging within the project.

I'll try splitting the build into two subprojects to see if it works with MSBuild.

**Reply by jatkinson1000** [reply]
Great @jalvesz - let us know how you get on.

I'll be honest that this is reaching the limits of my Windows knowledge, but would it be possible to build FTorch with NMake and install as a standalone library, then link your code to that with MSBuild?
You'd only be linking directly against the `.mod` file of the library...

If you want more debugging options you can try building with the `-DCMAKE_BUILD_TYPE=Debug` option as described in the README.md.

Are you bound to the Windows environment as we found the process smoother in WSL?

**Reply by jalvesz** [reply, marked as answer]
>Are you bound to the Windows environment as we found the process smoother in WSL?

This is the main issue, WSL is indeed useful, but at my place, our dev environment is windows first, linux second, I'm one of the few who actually uses WSL, but if I want my team to work with what I would like to test with your library, it has to build/run/debug natively on Windows first.

>I'll be honest that this is reaching the limits of my Windows knowledge, but would it be possible to build FTorch with NMake and install as a standalone library, then link your code to that with MSBuild?

I might come around this option later, but I found a way to tweak the CMakeList.txt in order to build with MSBuild. I duplicated a CMake on top of the folder containing FTorch (to avoid touching your original sources). Here just an extract of the section in which I split the library:
```CMake
...
# Library with C bindings
set(CLIB_NAME cftorch)
add_library(${CLIB_NAME} STATIC 
            FTorch/src/ctorch.cpp)
set_target_properties(${CLIB_NAME} PROPERTIES
  PUBLIC_HEADER "FTorch/src/ctorch.h"
  FOLDER "third_party/${PROJECT_NAME}"
)
# Link TorchScript
target_link_libraries(${CLIB_NAME} PRIVATE ${TORCH_LIBRARIES})
install(TARGETS ${CLIB_NAME})

# Library with Fortran bindings
add_library(${LIB_NAME} SHARED 
            FTorch/src/ftorch.f90)
# Add an alias FTorch::ftorch for the library
add_library(${PROJECT_NAME}::${LIB_NAME} ALIAS ${LIB_NAME})
set_target_properties(${LIB_NAME} PROPERTIES
  FOLDER "third_party/${PROJECT_NAME}"
)
target_link_libraries(${LIB_NAME} PRIVATE ${CLIB_NAME})
# Include the Fortran mod files in the library
target_include_directories(${LIB_NAME} PUBLIC )

# Install library, create target file
install(TARGETS ${LIB_NAME})
```
Managed to build the FTorch library and link another project against it :)
![image](https://github.com/Cambridge-ICCS/FTorch/assets/102541118/7d312701-8a48-4965-9105-5cda759d8916)

I'll keep you updated on my user experience with it

Thanks a lot!!

**Reply by jatkinson1000** [reply]
Ah, so you built as two separate libraries, one C and one Fortran, then link together.
That's a nice solution, thanks for updating us!
Please do let us know how you get on and how we can make the library better - and we are happy to feature any work/projects using it in our README.
